### PR TITLE
Core implementation

### PR DESCRIPTION
**Pull request by julien-thierry**
This is the implementation of the kpatch plugin for dnf.

I'm not sure about the proper use of license and thing like that so please point out any issues regarding that.

### CODE REVIEW

File: man/dnf.plugin.kpatch.8 [hunk 1]
```diff
@@ -0,0 +1,34 @@
+.TH man 8 "23 Apr 2020" "1.0" "dnf-kpatch man page"
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:1]: Should be `.TH "DNF-KPATCH" ...`.

Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:1]: File license/copyright information is missing.

Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:1]: I'd rather leave out `extra3` argument of `.TH` macro, as it is substituted by the section description.

File: man/dnf.plugin.kpatch.8 [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+.SH NAME
+dnf\-kpatch \- kpatch patch module installation plugin
+
+.SH SYNOPSIS
+dnf kpatch <install\-mode>
+
+.SH DESCRIPTION
+
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:9]: Extra newline?

File: man/dnf.plugin.kpatch.8 [hunk 3]
```diff
@@ -0,0 +1,34 @@
+.TH man 8 "23 Apr 2020" "1.0" "dnf-kpatch man page"
+.SH NAME
+dnf\-kpatch \- kpatch patch module installation plugin
+
+.SH SYNOPSIS
+dnf kpatch <install\-mode>
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:6]: What about somthing like this?
`.SY dnf`
`.B kpatch`
`.I install\-mode`

File: man/dnf.plugin.kpatch.8 [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+.SH DESCRIPTION
+
+dnf\-kpatch is a dnf plugin providing automated installation of kpatch\-patch
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:10]: `.B dnf\-kpatch` or \fBdnf\-kpatch\fR

Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:10]: `.BR dnf (8)`

File: man/dnf.plugin.kpatch.8 [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+packages. When enabled, installation of new kernel packages via dnf will trigger
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:11]: `.BR dnf (8)`

File: man/dnf.plugin.kpatch.8 [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+installation of the latest kpatch-patch package targeting the same kernel version.
+
+
+.SH INSTALL MODES
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:15]: `.SH OPTIONS`
`.TP`
`.B install\-mode`
...

File: man/dnf.plugin.kpatch.8 [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:16]: Extra newline.

File: man/dnf.plugin.kpatch.8 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+.B auto
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:17]: `.RS`
`.TP 8`
`.B auto`
...

File: man/dnf.plugin.kpatch.8 [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+       Installs missing kpatch\-patch packages for installed kernels and enables
+       automatic kpatch-patch installation for future kernel installations.
+
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:20]: Extra newline.

File: man/dnf.plugin.kpatch.8 [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+.B manual
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:21]: `.TP`
`.B manual`
...

File: man/dnf.plugin.kpatch.8 [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+       Disables automatic installation of kpatch-patch pacakges.
+
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:23]: `.RE`

File: man/dnf.plugin.kpatch.8 [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+.SH SEE ALSO
+dnf(1) kpatch (1)
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:26]: `.BR dnf (8),`
`.BR kpatch (1)`

File: man/dnf.plugin.kpatch.8 [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+.SH BUGS
+No known bugs.
+
+.SH AUTHOR
+Julien Thierry <jthierry@redhat.com>
+.SH COPYRIGHT
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:33]: Missing newline.

File: man/dnf.plugin.kpatch.8 [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+
+.SH SEE ALSO
+dnf(1) kpatch (1)
+
+.SH BUGS
+No known bugs.
+
+.SH AUTHOR
+Julien Thierry <jthierry@redhat.com>
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:32]: `.UR mailto:jthierry@redhat.com`
`Julien Thierry`
`.UE`

File: man/dnf.plugin.kpatch.8 [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+.B manual
+       Disables automatic installation of kpatch-patch pacakges.
+
+
+.SH SEE ALSO
+dnf(1) kpatch (1)
+
+.SH BUGS
+No known bugs.
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:29]: Missing "REPORTING BUGS" section.

File: man/dnf.plugin.kpatch.8 [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+.SH AUTHOR
+Julien Thierry <jthierry@redhat.com>
+.SH COPYRIGHT
+Copyright (C) 2020: Julien Thierry <jthierry@redhat.com>
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:34]: Missing program license information.

Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:34]: Extra colon.

File: man/dnf.plugin.kpatch.8 [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+dnf\-kpatch \- kpatch patch module installation plugin
+
+.SH SYNOPSIS
+dnf kpatch <install\-mode>
+
+.SH DESCRIPTION
+
+dnf\-kpatch is a dnf plugin providing automated installation of kpatch\-patch
+packages. When enabled, installation of new kernel packages via dnf will trigger
+installation of the latest kpatch-patch package targeting the same kernel version.
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:12]: `kpatch\-patch`

**Comment by julien-thierry** [comment]
> The manual page reads okay and the code looks reasonable, though I'm not too familiar with python or the plugin API :) Can you provide a quick summary of the testing that was performed on the plugin. Are there any weird corner cases like kernels w/o associated kpatch-patches, or custom kernel rpms without .el7/8 naming conventions that might throw the plugin off? Thanks.

In terms of functionality, I tested installing a couple of kernels, enabling the plugin, checking the expected kpatch-patch packages would be added for installation. Installing another kernel, checking the corresponding kpatch-patch package is added for installation. Then disabling the plugin, installing another kernel, making sure kpatch-patch packages are no longer pulled in. Then re-enabling the plugin to check the missing kpatch-patch would get installed.

I admit it's not the most exhaustive testing but I'm open to suggestions.

Regarding naming conventions, I'm not sure how flexible we want to be. For now, only RHEL-8 is supported, and assumes that the kernel kpatch fixes are delivered with the name `kpatch-patch-${kernel_version/./_}-${kernel_release/./_}`. For now it does handle ".el.*" rpm extentions. I'll happily make it more generic, but it would be good to know how other dnf based distros decide to proceed.

For kernel w/o associated kpatch-patches, if no packages are found the plugin can't add them and simply does nothing. I tested it by disabling the repository containing the kpatch-patch packages and enabling the plugin.

**Comment by julien-thierry** [comment]
@esyr-rh Thanks for the suggestions on the man page, I'll act on them and take the opportunity to update the kpatch and kpatch-build man pages following your formatting suggestions.

**Comment by joe-lawrence** [comment]
> Regarding naming conventions, I'm not sure how flexible we want to be. For now, only RHEL-8 is supported, and assumes that the kernel kpatch fixes are delivered with the name `kpatch-patch-${kernel_version/./_}-${kernel_release/./_}`. For now it does handle ".el.*" rpm extentions. I'll happily make it more generic, but it would be good to know how other dnf based distros decide to proceed.

I didn't mean to suggest supporting other naming conventions, I was only wondering if kernels without .el* would break the plugin.

> For kernel w/o associated kpatch-patches, if no packages are found the plugin can't add them and simply does nothing. I tested it by disabling the repository containing the kpatch-patch packages and enabling the plugin.

... to which I think this would cover any strangely named kernels as well.

File: man/dnf.plugin.kpatch.8 (hunk 1 above)
Comment (julien-thierry) [reply man/dnf.plugin.kpatch.8]: Is the license added at the bottom of the file enough?

File: man/dnf.plugin.kpatch.8 (hunk 14 above)
Comment (julien-thierry) [reply man/dnf.plugin.kpatch.8]: This gives a weird output in the man:
`Julien Thierry (mailto:jthierry@redhat.com)`

Is that intended?

File: man/dnf.plugin.kpatch.8 (hunk 1 above)
Comment (esyr-rh) [reply man/dnf.plugin.kpatch.8]: It depends; generally, man pages in some projects may have different license (CC-like, for example), and some projects insist on having license (or SPDX tag) in each file. If kpatch is not one of those, then one may assume that the project's license applies to the man page as well, yes.

File: man/dnf.plugin.kpatch.8 (hunk 14 above)
Comment (esyr-rh) [reply man/dnf.plugin.kpatch.8]: Yes; it's, however, a matter of personal preference (unless there's a project-wide policy in place) on how to format (e-mail) links; to my taste, `.UR` macro is more "semantic".

Comment (julien-thierry) [reply man/dnf.plugin.kpatch.8]: I see, I'll keep your suggestion and see if the rest of the team is happy with it or has other taste :)

File: man/dnf.plugin.kpatch.8 (hunk 1 above)
Comment (julien-thierry) [reply man/dnf.plugin.kpatch.8]: I think the project's license applies to the man pages. @jpoimboe would you confirm that? Or is there someone else to discuss this?

File: man/dnf.plugin.kpatch.8 [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+.SH NAME
+dnf\-kpatch \- kpatch patch module installation plugin
+
+.SH SYNOPSIS
+.SY dnf
+.B kpatch
+.I <install\-mode>
+
+.SH DESCRIPTION
+\fBdnf\-kpatch\fR is a \fBdnf(8)\fR plugin providing automated installation of
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:11]: Section number is not emboldened, like this: `\fBdnf\fR(8)`.

File: man/dnf.plugin.kpatch.8 [hunk 19]
```diff
@@ -0,0 +1,50 @@
+.TH "DNF-KPATCH" 8
+.SH NAME
+dnf\-kpatch \- kpatch patch module installation plugin
+
+.SH SYNOPSIS
+.SY dnf
+.B kpatch
+.I <install\-mode>
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:8]: There's no need in angle brackets.

File: man/dnf.plugin.kpatch.8 [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+.SH DESCRIPTION
+\fBdnf\-kpatch\fR is a \fBdnf(8)\fR plugin providing automated installation of
+kpatch\-patch packages. When enabled, installation of new kernel packages via
+\fBdnf(8)\fR will trigger installation of the latest kpatch-patch package
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:13]: `\fBdnf\fR(8)`

File: man/dnf.plugin.kpatch.8 [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+Installs missing kpatch\-patch packages for installed kernels and enables
+automatic kpatch\-patch installation for future kernel installations.
+.TP
+.B manual
+Disables automatic installation of kpatch-patch pacakges.
+.RE
+
+.SH SEE ALSO
+.BR dnf(8) ,
+.BR kpatch(1)
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:31]: `.BR dnf (8),`
`.BR kpatch (1)`

**Comment by julien-thierry** [comment]
> ... to which I think this would cover any strangely named kernels as well.

Yes, if the packages have unexpected names, the plugin just wouldn't see them/look for them and not to anything with respect to those packages.
In theory, the plugin shouldn't "block" things in dnf, preventing installation or removal of packages, nor cause unexpected packages to be installed. If that's not the case, it'll need fixing :) .

File: man/dnf.plugin.kpatch.8 [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+kpatch\-dnf project page
+.UE .
+
+.SH AUTHOR
+.UR mailto:jthierry@redhat.com
+Julien Thierry
+.UE
+
+.SH COPYRIGHT
+Copyright (C) 2020 Julien Thierry <jthierry@redhat.com>
```
Comment (esyr-rh) [inline man/dnf.plugin.kpatch.8:48]: It's possible to use `\[co]` for the copyright symbol (see `groff_char(7)`).

Comment (julien-thierry) [reply man/dnf.plugin.kpatch.8]: Thanks.

File: man/dnf.plugin.kpatch.8 (hunk 1 above)
Comment (jpoimboe) [reply man/dnf.plugin.kpatch.8]: Yes, the project license applies to everything, including the man pages.

**Comment by jpoimboe** [comment]
Looks good to me, just a few comments about naming.

According to the man page the plugin is called "kpatch", which I think is the right name.  So I think KPATCH_PLUGIN_NAME should be "kpatch"?

Also for consistency with other dnf plugins, and considering the fact that the files will probably be installed in appropriately named subdirs, I think the files should be named kpatch.conf, dnf-kpatch.8, and kpatch.py.

**Comment by julien-thierry** [comment]
@jpoimboe 

> According to the man page the plugin is called "kpatch", which I think is the right name. So I think KPATCH_PLUGIN_NAME should be "kpatch"?
> 
So, my main worry here was about having a name conflict/confusion somewhere with the kpatch package/executable itself. It might be completely unjustified just wanted to avoid potential issues. But I can do the renaming.

> Also for consistency with other dnf plugins, and considering the fact that the files will probably be installed in appropriately named subdirs, I think the files should be named kpatch.conf, dnf-kpatch.8, and kpatch.py.

Regarding the man page naming, on RHEL8 I get the following:
```
$ ls -1 /usr/share/man/man8/dnf*
/usr/share/man/man8/dnf.8.gz
/usr/share/man/man8/dnf.plugin.builddep.8.gz
/usr/share/man/man8/dnf.plugin.changelog.8.gz
/usr/share/man/man8/dnf.plugin.config_manager.8.gz
/usr/share/man/man8/dnf.plugin.copr.8.gz
/usr/share/man/man8/dnf.plugin.debug.8.gz
/usr/share/man/man8/dnf.plugin.debuginfo-install.8.gz
/usr/share/man/man8/dnf.plugin.download.8.gz
/usr/share/man/man8/dnf.plugin.generate_completion_cache.8.gz
/usr/share/man/man8/dnf.plugin.needs_restarting.8.gz
[...]
```

But on Fedora I get:

```
$ ls -1  /usr/share/man/man8/dnf*
/usr/share/man/man8/dnf.8.gz
/usr/share/man/man8/dnf-builddep.8.gz
/usr/share/man/man8/dnf-changelog.8.gz
/usr/share/man/man8/dnf-config-manager.8.gz
/usr/share/man/man8/dnf-copr.8.gz
/usr/share/man/man8/dnf-debug.8.gz
/usr/share/man/man8/dnf-debuginfo-install.8.gz
/usr/share/man/man8/dnf-download.8.gz
/usr/share/man/man8/dnf-generate_completion_cache.8.gz
/usr/share/man/man8/dnf-needs-restarting.8.gz
```

I followed the RHEL-8 convention as it is the only distro supported for now. But in the end the file can be renamed when packaging it into a rpm to fit the disto convention.

So, just let me know which name we'd prefer upstream.

**Comment by esyr-rh** [comment]
On Thu, Apr 30, 2020 at 12:10:43AM -0700, julien-thierry wrote:
> Regarding the man page naming, on RHEL8 I get the following:
> ```
> $ ls -1 /usr/share/man/man8/dnf*
> /usr/share/man/man8/dnf.8.gz
> /usr/share/man/man8/dnf.plugin.builddep.8.gz
> /usr/share/man/man8/dnf.plugin.changelog.8.gz
> /usr/share/man/man8/dnf.plugin.config_manager.8.gz
> /usr/share/man/man8/dnf.plugin.copr.8.gz
> /usr/share/man/man8/dnf.plugin.debug.8.gz
> /usr/share/man/man8/dnf.plugin.debuginfo-install.8.gz
> /usr/share/man/man8/dnf.plugin.download.8.gz
> /usr/share/man/man8/dnf.plugin.generate_completion_cache.8.gz
> /usr/share/man/man8/dnf.plugin.needs_restarting.8.gz
> [...]

FWIW, RHEL 8 naming doesn't make any sense, as man pages (in section 8,
at least) are named after commands and not some internal constructs.

**Comment by julien-thierry** [comment]
> FWIW, RHEL 8 naming doesn't make any sense, as man pages (in section 8, at least) are named after commands and not some internal constructs.

Yes, I had that thought as well but went with it's convention anyway.

I guess I'll use the more logical convention (without the .plugin.) and we'll rename it when packaging it for RHEL-8 in the rpm spec file.

**Comment by esyr-rh** [comment]
On Thu, Apr 30, 2020 at 02:34:08AM -0700, julien-thierry wrote:
> > FWIW, RHEL 8 naming doesn't make any sense, as man pages (in section 8, at least) are named after commands and not some internal constructs.
> 
> Yes, I had that thought as well but went with it's convention anyway.
> 
> I guess I'll use the more logical convention (without the .plugin.) and we'll rename it when packaging it for RHEL-8 in the rpm spec file.

Note also that it seems that it has been changed to "dnf-subcommand.8.gz"
in RHEL 8.2.0.

**Comment by jpoimboe** [comment]
I think it needs a license.  dnf itself seems to be GPLv2+ so that might be a good option.

**Comment by julien-thierry** [comment]
> I think it needs a license. dnf itself seems to be GPLv2+ so that might be a good option.

I've added this with the last update.

**Comment by julien-thierry** [comment]
I'd like to merge this to have a first base version.
Any objections?

**Review by sm00th — APPROVED** [review]

**Pull request merged**